### PR TITLE
MB-10533 Locust best practices refactors

### PR DESCRIPTION
**Pull request by felipe-lee**
## Description

Implementing some of my suggestions for best practices refactors. Key points are:
* Use a `RestTaskSet` class that can handle automatically parsing json and will auto-fail if the response isn't json. 
    * This also gives us the ability to mark a task as a success or failure, regardless of the status code.
    * Note that this is different from the original plan to use `RestHttpUser` because we don't really need the things it provides at a user level, we need them at a task set level since we define all our load tests in task sets.
* Make it easier to make requests with all the necessary bits (proper url, certs, etc.). To do this, I've created some helper functions that give you what you need, which can then be used with the `client`, or with our own thing, e.g. we can use the `requests` library directly.
* Create some parent classes we can easily use without having to worry about the proper order for mixins to be in.
    * Only have one so far, but I'm going to take a look to see if anything else would be good to do. 
* Pull cert creation/deletion out of the user classes and instead do it at the locust init/shutdown level. This simplifies it a bit. Also cut down on the ability to run against arbitrary environments since we are only using local and dp3.

## Setup

Things are still in progress, but I added a locustfile called `rest.py` that shows a very simple usage of the new `RestHttpUser` class. You can run this file by starting up the mymove server, and then running this command:

```sh
locust -f locustfiles/rest.py --host local
```

If you'd like to run without the UI, here's the command with some extra args:

```sh
locust -f locustfiles/rest.py --headless --host local -u 100 -r 5 -t 30s -T listMoves
```

## Code Review Verification Steps

* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-10533) for this change.
* [Locust Best Practices Wiki](https://dp3.atlassian.net/wiki/spaces/MT/pages/1582891038/Locust+Best+Practices)
    * I need to update this wiki though with some of my learnings as I've tried to implement changes...

### CODE REVIEW

File: locustfiles/lifecycle.py
Comment (duncan-truss) [inline locustfiles/lifecycle.py:6]: It looks like both our repo and theirs are MIT licensed, maybe we should include their copyright as well in this file https://github.com/locustio/locust/blob/master/LICENSE

Comment (felipe-lee) [reply locustfiles/lifecycle.py:6]: Tbh, I don't know much about the license stuff. Do you mean like include a link to their license? Or add a copyright statement?

Comment (duncan-truss) [reply locustfiles/lifecycle.py:6]: I would say add the copyright statement to the top of that file if that's what you copied here.

File: utils/hosts.py
Comment (ronaktruss) [inline utils/hosts.py:56]: In the previous iteration, `is_api` was only true for the prime tests. So when running the office tests we wouldn't expect the `api` prefix as part of the host name. Is this still the case/ is there a related change that modifies the way this function is used.

Comment (felipe-lee) [reply utils/hosts.py:56]: 😞 I swear I had seen it on the other ones as well, but now that you called it out and I'm looking on master, you're right. I'll bring this back in. That's what I get for primarily only testing the prime api with my changes. Thanks for catching this!

Comment (felipe-lee) [reply utils/hosts.py]: I kind of reverted it. I didn't do a direct revert because `is_api` is somewhat confusing since they're all APIs. 

Here is the commit that brings this back: https://github.com/transcom/milmove_load_testing/pull/97/commits/3bd1f6624afa560c4b619eff358f4f62549bfb73

Though I am also thinking of maybe changing how URLs are formed to be this way: https://github.com/transcom/milmove_load_testing/pull/97/commits/8196f38fa91ce6f8a43af316b3cba2400a12cb22

File: tasks/prime.py
Comment (ronaktruss) [inline tasks/prime.py:925]: Extra `<` character in the path

**Review by ronaktruss — COMMENTED** [review]
LGTM. I think there are areas that are a bit confusing at the moment since we are supporting both the old way and the new way. Specifically the way that urls are setup in the MilMoveHostMixin, which to my understanding isn't used in the new format you are proposing.

File: tasks/prime.py
Comment (felipe-lee) [reply tasks/prime.py:925]: Oh good catch, thanks!

**Comment by felipe-lee** [comment]
> LGTM. I think there are areas that are a bit confusing at the moment since we are supporting both the old way and the new way. Specifically the way that urls are setup in the MilMoveHostMixin, which to my understanding isn't used in the new format you are proposing.

Yeah, I'm going to see how much work it would be to just switch things over so we don't have to live in both worlds. Based on my earlier testing, it should hopefully not be too much 🤞🏼

**Comment by felipe-lee** [comment]
I've written tests for all the helpers and such I wrote. At least I think I have... 🤔 ...but it's also the end of the day so I may have missed something. At this point, the main thing left is to address what @ronaktruss pointed out about it being confusing at the moment because we have both the new and old ways of doing things. I'll tackle that on monday, and depending on how much it is, might open a second PR to stop this one from getting too much bigger. We'll see. For now though, the core things that I've added I think are ready for review.

File: utils/auth.py
Comment (duncan-truss) [inline utils/auth.py:35]: Does hardcoding this prevent us from running against say experimental?

Comment (felipe-lee) [reply utils/auth.py:35]: To be honest, I'm not sure. In the past I would have said yes, but since experimental got moved to `dp3`, it might be that our cert key might work with either one. But it would require us to edit other code like the bit about which base domain to point to.

I wasn't too concerned since my understanding is that we really only want to run against the load testing environment. We only used experimental before because we didn't have a dedicated environment. Do you think we should keep the ability to point at other environments though?

**Comment by felipe-lee** [comment]
Note that I'm still working through updating the README, but as I was doing code examples for it, I realized we really don't need the `RestHttpUser` class, but rather that functionality at a `TaskSet` level since we define all our load tests in task sets.

File: utils/auth.py
Comment (duncan-truss) [reply utils/auth.py:35]: Oh I had forgotten that exp moved outside of the ATO boundary, I was mainly thinking we'd want to know if load testing performs differently inside the ATO infra.

It's a bit weird that we expose the host field in locust as configurable/overideable but make slightly different assumptions here.

No action needed.

Comment (felipe-lee) [reply utils/auth.py:35]: Yeah that is one weird change this introduces in that we don't really use the locust host for much other than as a starting point. Normally the locust host is used in the background by the locust client to create the full URL, but these changes make it so that the host is just a base string that we start off with and then we pass the client full urls that our `MilMoveRequestPreparer` sets up

File: README.md
Comment (duncan-truss) [inline README.md:76]: It might be nice to say that we are only really testing the APIs and backend of the system.  It's not giving us insight into anything browser or client side.

File: README.md
Comment (duncan-truss) [inline README.md:103]: just for consistency this is the only folder without a trailing slash `scripts/`

File: README.md
Comment (felipe-lee) [reply README.md:76]: Ah good point, I'll add that.

File: README.md
Comment (felipe-lee) [reply README.md:103]: oh good catch, thanks!

File: README.md
Comment (duncan-truss) [inline README.md:168]: I think this is true the first time but I was generally having to run `pipenv install` unless I deactivated or destroyed the virtual environment.

Comment (felipe-lee) [reply README.md]: Ah interesting, yeah it looks like direnv only runs the install if there isn't already a venv: 
https://github.com/direnv/direnv/blob/f68f741895754e9551059c8e0132a796ad1c3c3a/stdlib.sh#L882-L904

I've updated the README with new instructions for updating python dependencies. Thanks for pointing this out!

File: locustfiles/lifecycle.py
Comment (felipe-lee) [reply locustfiles/lifecycle.py:83]: Me either, but it's pretty cool for testing things.

File: utils/tests/test_users.py
Comment (felipe-lee) [reply utils/tests/test_users.py]: I think this should be gone after this commit: https://github.com/transcom/milmove_load_testing/pull/97/commits/1c05f0d53cbdc31f7deabb360d681818394ac48d

File: README.md
Comment (duncan-truss) [inline README.md:308]: I might spell out loadtest.dp3.us here, I think it gets a little muddled with all the environment, host, and server terms being thrown around getting used.

Comment (felipe-lee) [reply README.md]: Does https://github.com/transcom/milmove_load_testing/pull/97/commits/1c32c1dd7eca887b8691dc4dce43b01ebf66f8fa help? Or do you mean some other way?

Comment (duncan-truss) [reply README.md]: Thanks that's exactly what I meant.

**Comment by felipe-lee** [comment]
Ok, the docs are done I think, at least based on where the code is now. I'll work on seeing what it takes to switch existing load tests over to use the new stuff in a separate PR to keep it more manageable.

File: .editorconfig
Comment (ronaktruss) [inline .editorconfig:12]: Is there any additional setup that we need to do to get this to work? All of the other rules seem to be working fine on my end other than this rule.

Comment (felipe-lee) [reply .editorconfig:12]: I don't think so, though maybe I should move it to just be for Markdown. I would like this to be set for python but didn't set it in `black`'s settings because it would have led to a lot of noise in this pr. Is it not working for any files for you? Or do you mean it isn't working for python files?

Comment (ronaktruss) [reply .editorconfig:12]: Yeah I saw the details of your commit msg indicating the reason for not including in the `black` settings. For me its not working in the markdown files for some reason. (I was trying to do a local diff since the current one on github is hard to parse and wasn't able to get it to work.)

Comment (felipe-lee) [reply .editorconfig:12]: Weird, I wonder why. You use VS code right?

Comment (felipe-lee) [reply .editorconfig:12]: Oh I see, apparently it isn't implemented in the VS code plugin because they don't think it should be part of it: https://github.com/editorconfig/editorconfig-vscode/issues/53

Comment (felipe-lee) [reply .editorconfig:12]: Hmm I guess maybe I should remove it completely then, otherwise we'll keep having differences between people using vscode and people using jetbrains IDEs

Comment (ronaktruss) [reply .editorconfig:12]: Yeah, I'm using VSCode. What's the rationale behind having max line lengths for md files? I've generally been a fan of using wrapping as it makes the diffs cleaner imo.

**Review by ronaktruss — APPROVED** [review]
LGTM. I reviewed it with a slightly wider tooth comb than usual (given the length and having looked at this during the drafts), so I'd recommend waiting for a second approval if possible.

File: .editorconfig
Comment (felipe-lee) [reply .editorconfig:12]: Mainly being able to have multiple things up side by side. I had the plain README, the preview, and a code window all side by side to help as I was writing and with the original super long lines it was harder to edit. Wrapping sometimes works weirdly, but I can change to using other screens or something.

File: utils/hosts.py
Comment (duncan-truss) [inline utils/hosts.py:52]: I think you have a `milmove_env.is_local()` method you can call here?

File: utils/request.py
Comment (duncan-truss) [inline utils/request.py:219]: This is right but could be overridden through the host form field if they start a new load test through the UI.

Comment (felipe-lee) [reply utils/request.py:219]: Ah good point, I'll update this.

File: utils/hosts.py
Comment (felipe-lee) [reply utils/hosts.py:52]: Ah, thanks for catching that

**Review by duncan-truss — APPROVED** [review]
New patterns look good to me, thanks for taking this on.

I ran into some issues on the dp3 load test environment but we can look into those after this is merged.

**Pull request merged**
